homeassistant/components/lifx/coordinator.py

```diff
@@ -188,42 +196,67 @@ def async_get_entity_id(self, platform: Platform, key: str) -> str | None:
 
     async def _async_update_data(self) -> None:
         """Fetch all device data from the api."""
-        async with self.lock:
-            if self.device.host_firmware_version is None:
-                self.device.get_hostfirmware()
-            if self.device.product is None:
-                self.device.get_version()
-            if self.device.group is None:
-                self.device.get_group()
-
-            response = await async_execute_lifx(self.device.get_color)
-
-            if self.device.product is None:
-                raise UpdateFailed(
-                    f"Failed to fetch get version from device: {self.device.ip_addr}"
+
+        for _ in range(MAX_TIMEOUTS_TO_DECLARE_UPDATE_FAILED - 1):
+            try:
+                tasks: list[Awaitable] = [async_execute_lifx(self.device.get_color)]
+
+                if self.device.host_firmware_version is None:
+                    tasks.append(async_execute_lifx(self.device.get_hostfirmware))
+                if self.device.product is None:
+                    tasks.append(async_execute_lifx(self.device.get_version))
+                if self.device.group is None:
+                    tasks.append(async_execute_lifx(self.device.get_group))
+
+                await asyncio.gather(*tasks)
+
+                events = []
+                # Make a copy of the bulb's message queue to avoid the dictionary changing
+                # size during iteration
+                for _response, event, _callb in list(self.device.message.values()):
+                    if isinstance(event, asyncio.Event):
+                        events.append(event.wait())
```

Comment:
```suggestion
                for _response, event, _callb in self.device.message.values():
                    if isinstance(event, asyncio.Event):
                        events.append(event.wait())
```

Sorry, we can drop the copy as we don't need it given this is all in the same thread.

I've been working too much with threaded code this week so my example was overly cautious